File: 算法/Python/第2章 分治法/2.2 大整数加法/big_int_add.py
def big_int_add(num1, num2):
    """
    大整数相加
    Args:
        num1 (str): 大整数1
        num2 (str): 大整数2
    Returns:
        [str]: 相加结果
    """
    # 其中一个数为0，直接返回另一个数
    if num1 == "0":
        return num2
    elif num2 == "0":
        return num1
    
    # 计算两个数中较长的整数位数
    max_len = max(len(num1), len(num2))
    # 让位数较短的整数前面补0对齐
    num1 = '0' * (max_len - len(num1)) + num1
    num2 = '0' * (max_len - len(num2)) + num2

    result = ""         # 结果
    carry = 0           # 保存进位
    # 从右往左逐位相加
    for i in range(max_len - 1, -1, -1):
        s = int(num1[i]) + int(num2[i]) + carry
        result = str(s % 10) + result
        carry = s // 10
    
    # 判断最高位是否有进位
    if carry > 0:
        result = str(carry) + result
    
    # 去除结果前面多余的0
    i = 0
    while result[i] == '0':
        i += 1
    return result[i:]

def main():
    print(big_int_add("426709752318", "95481253129"))

if __name__ == "__main__":
    main()

File: 算法/Java/第2章 分治法/2.2 大整数加法/BigIntAdd.java
import java.util.Collections;

public class BigIntAdd {
    public static void main(String[] args) {
        System.out.println(bigIntAdd("426709752318", "95481253129"));
    }

    /**
	 * @brief   大整数加法
	 * @param   num1: 大整数1
	 * @param   num2: 大整数2
	 * @retval  相加结果
	 */
    public static String bigIntAdd(String num1, String num2) {
        // 其中一个数为0，直接返回另一个数
    	if(num1.equals("0")) {
            return num2;
        } else if(num2.equals("0")) {
            return num1;
        }
    	
    	// 计算两个数中较长的整数位数
    	int maxLen = Math.max(num1.length(), num2.length());
    	// 让位数较短的整数前面补0对齐
    	num1 = String.join("", Collections.nCopies(maxLen - num1.length(), "0")) + num1;
    	num2 = String.join("", Collections.nCopies(maxLen - num2.length(), "0")) + num2;
    	
    	String result = "";		// 结果
    	int carry = 0;			// 保存进位
    	// 从右往左逐位相加
    	for(int i = maxLen - 1; i >= 0; i--) {
    		int s = (num1.charAt(i) - '0') + (num2.charAt(i) - '0') + carry;
    		result = s % 10 + result;
    		carry = s / 10;
    	}
    	
    	// 判断最高位是否有进位
    	if(carry > 0) {
    		result = carry + result;
    	}
    	
    	// 去除结果前面多余的0
    	int i = 0;
    	while(result.charAt(i) == '0') {
    		i++;
    	}
        return result.substring(i);
    }
}